### PR TITLE
REF: de-duplicate Block.__init__

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/internals/blocks.py [hunk 1]
```diff
@@ -134,16 +134,20 @@ def __init__(self, values, placement, ndim: int):
...
 
         if self._validate_ndim and self.ndim and len(self.mgr_locs) != len(self.values):
             raise ValueError(
                 f"Wrong number of items passed {len(self.values)}, "
                 f"placement implies {len(self.mgr_locs)}"
             )
 
+        if self.is_extension and self.ndim == 2 and len(self.mgr_locs) != 1:
+            # TODO(EA2D): check unnecessary with 2D EAs
+            raise AssertionError("block.size != values.size")
```
Comment (jreback) [inline pandas/core/internals/blocks.py:149]: test hit this?

File: pandas/core/internals/blocks.py [hunk 2]
```diff
@@ -180,7 +184,19 @@ def _check_ndim(self, values, ndim):
         ValueError : the number of dimensions do not match
         """
         if ndim is None:
-            ndim = values.ndim
+            warnings.warn(
+                "Accepting ndim=None in the Block constructor is deprecated, "
```
Comment (jreback) [inline pandas/core/internals/blocks.py:188]: ahh i c, can you open an issue on arrow jira tracker and link to this PR for updating this

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/internals/blocks.py [hunk 3]
```diff
@@ -180,7 +184,19 @@ def _check_ndim(self, values, ndim):
...
+                "this will raise in a future version.",
+                FutureWarning,
```
Comment (jreback) [inline pandas/core/internals/blocks.py:190]: normally i want FutureWarning, but in this case this is only an internal library use. ok to use DeprecationWarning here

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
ok for 1.2

**Comment by jreback** [comment]
@jbrockmendel if you can address these comments can get this in

**Comment by jbrockmendel** [comment]
xref https://issues.apache.org/jira/browse/ARROW-10768

**Review by jorisvandenbossche — CHANGES_REQUESTED** [review]
What's the rationale for this change? Why is it needed to deprecate it? (pyarrow code currently works fine? Or is there a problem with it)

**Comment by jbrockmendel** [comment]
> What's the rationale for this change? Why is it needed to deprecate it? (pyarrow code currently works fine? Or is there a problem with it)

We recently made it so that all our internal calls explicitly pass ndim.  Now we're telling other packages (that ideally wouldnt be accessing internals in the first place) they need to do the same.

**Comment by jorisvandenbossche** [comment]
> they need to do the same.

But so is there an actual reason to require this, apart from "we are not using the keyword anymore internally" ? 
(I didn't closely follow your prior work related to `ndim`, so I am honestly asking. Does it give risks on bugs for the external users, or is it slower, or ... when they don't explicitly pass the `ndim`?)

If we do the change, we could also keep this for after the 1.2 RC, so the deprecation is only for 1.3, which gives pyarrow time to fix the code and have a release before this is released by pandas.

**Comment by jbrockmendel** [comment]
> But so is there an actual reason to require this, apart from "we are not using the keyword anymore internally" ?
(I didn't closely follow your prior work related to ndim, so I am honestly asking. Does it give risks on bugs for the external users, or is it slower, or is it hard to maintain that functionality for us, or ... when they don't explicitly pass the ndim?)

The motivation that led me to update all of our internal usages to always pass ndim explicitly was that on other branches I was getting errors related to incorrect ndim inference.  (Keep in mind I have active branches for both ArrayManager and EA2D among others and I don't recall off the top of my head which one(s) were producing these errors)

> or is it slower

ndim inference isn't going to be particularly slow, but we do call the Block constructors a lot, so it _could_ add up.

> or is it hard to maintain that functionality for us

In general I think being more explicit and more strict in internals code makes maintenance easier, yes.

**Comment by simonjayhawkins** [comment]
> If we do the change, we could also keep this for after the 1.2 RC, so the deprecation is only for 1.3, which gives pyarrow time to fix the code and have a release before this is released by pandas.

@jbrockmendel can we remove the blocker tag? and do this for 1.3?

**Comment by jbrockmendel** [comment]
> and do this for 1.3?

sounds good

**Comment by jbrockmendel** [comment]
Mothballing pending https://github.com/apache/arrow/pull/8957

**Comment by jbrockmendel** [comment]
re-opening; now that make_block is exposed specifically for downstream, we can do the validation there and be stricter in private-internals

File: pandas/core/internals/blocks.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/internals/blocks.py:175]: yes, test_invalid_ea_block gets here

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm @jorisvandenbossche if any issues.

**Comment by jorisvandenbossche** [comment]
To make it explicit: the PR now no longer includes a deprecation warning. But you still changed the behaviour of `Block.__init__`, or not?

**Comment by jbrockmendel** [comment]
> To make it explicit: the PR now no longer includes a deprecation warning. But you still changed the behaviour of Block.__init__, or not?

`Block.__init__` is changed, but the relevant validation is now done in `api.make_block`, so downstream libraries using that will be unaffected.

**Comment by jreback** [comment]
> Yep, but thus downstream libraries using `Block` can be affected. But since we are not aware of such libraries, and the known ones use `make_block` / `make_block_same_class`, that's probably OK

great

**Comment by jreback** [comment]
thanks @jbrockmendel

**Comment by jbrockmendel** [comment]
alright! between this, #40262, and a couple follow-ups to this, im hopeful to take another big chunk out of the difference between libreduction-vs-purepython paths (xref #40263)

**Pull request merged**